registry/consul/listener.go

```diff
@@ -0,0 +1,196 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package consul
+
+import (
+	"sync"
+)
+
+import (
+	consul "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/api/watch"
+	perrors "github.com/pkg/errors"
+)
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/registry"
+	"github.com/apache/dubbo-go/remoting"
+)
+
+// Consul listener wraps the consul watcher, to
+// listen the service information change in consul
+// registry.
+type consulListener struct {
+	// Registry url.
+	registryUrl common.URL
+
+	// Consumer url.
+	consumerUrl common.URL
+
+	// Consul watcher.
+	plan *watch.Plan
+
+	// Most recent service urls return by
+	// watcher.
+	urls []common.URL
+
+	// All service information changes will
+	// be wrapped into ServiceEvent, and be
+	// sent into eventCh. Then listener's
+	// Next method will get event from eventCh,
+	// and return to upstream.
+	eventCh chan *registry.ServiceEvent
+
+	// All errors, happening in the listening
+	// period, will be caught and send into
+	// errCh. Then listener's Next method will
+	// get error from errCh, and return to
+	// upstream.
+	errCh chan error
+
+	// Running field represents whether consul
+	// listener has been closed. When closing
+	// listener, this field will be set to
+	// false, and will notify consul watcher
+	// to close.
+	running bool
+
+	// After listener notifies consul watcher
+	// to close, listener will call wg.wait to
+	// make sure that consul watcher is closed
+	// before the listener closes.
+	wg sync.WaitGroup
+}
+
+func newConsulListener(registryUrl common.URL, consumerUrl common.URL) (*consulListener, error) {
+	params := make(map[string]interface{})
+	params["type"] = "service"
+	params["service"] = consumerUrl.Service()
+	params["tag"] = "dubbo"
+	params["passingonly"] = true
+	plan, err := watch.Parse(params)
+	if err != nil {
+		return nil, err
+	}
+
+	listener := &consulListener{
+		registryUrl: registryUrl,
+		consumerUrl: consumerUrl,
+		plan:        plan,
+		urls:        make([]common.URL, 0),
+		eventCh:     make(chan *registry.ServiceEvent, 1),
+		errCh:       make(chan error, 1),
+		running:     true,
+	}
+
+	// Set handler to consul watcher, and
+	// make watcher begin to watch service
+	// information change.
+	listener.plan.Handler = listener.handler
+	listener.wg.Add(1)
+	go listener.run()
+	return listener, nil
+}
+
+// Wrap the consul watcher run api. There are three
+// conditions that will finish the run:
+//   - set running to false
+//   - call plan.Stop
+//   - close eventCh and errCh
+// If run meets first two conditions, it will close
+// gracefully. However, if run meets the last condition,
+// run will close with panic, so use recover to cover
+// this case.
+func (l *consulListener) run() {
+	defer l.wg.Done()
+	defer func() { recover() }()
```

Comment:
Why do you recover a `panic`? And no any handler.

Reply:
1. This is because when I close the listener, I will close the channel:

https://github.com/apache/dubbo-go/blob/eeb117760d397cd37e6832585fca4219b81fffcc/registry/consul/listener.go#L193-L194

and `run` function is in another goroutine, and may still send thing into channel:

https://github.com/apache/dubbo-go/blob/eeb117760d397cd37e6832585fca4219b81fffcc/registry/consul/listener.go#L176-L178

This will raise a panic in the `run` function. 

2. I have added a comment above the `run` function to explain why I need to recover a `panic`: 

https://github.com/apache/dubbo-go/blob/eeb117760d397cd37e6832585fca4219b81fffcc/registry/consul/listener.go#L111-L120 

3. I don't know how to specify a handler to handle this situation, so I use a `recover` to handle every panic that will happen in `run` function.

Reply:
Get it.  I think a error need to be output through `logger.Warn`.

Reply:
@fangyincheng Oh, yes. Adding a `logger.Warn` to print error will be better.